src/main/java/com/google/firebase/internal/FirebaseScheduledExecutor.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.internal;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.google.common.base.Strings;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.ThreadFactory;
+
+/**
+ * A single-threaded scheduled executor implementation. Allows naming the threads, and spawns
+ * new threads as daemons.
+ */
+public class FirebaseScheduledExecutor extends ScheduledThreadPoolExecutor {
+
+  public FirebaseScheduledExecutor(ThreadFactory threadFactory, String name) {
+    this(threadFactory, name, null);
+  }
+
+  public FirebaseScheduledExecutor(
+      ThreadFactory threadFactory, String name, Thread.UncaughtExceptionHandler handler) {
+    super(1, decorateThreadFactory(threadFactory, name, handler));
```

Comment:
I think it would be super duper awesome if we just used `ScheduledThreadPoolExecutor` in our code. It would be much easier to grasp that we are now using an off-the-shelf executor. Maybe you can find a different place for `decorateThreadFactory`?

Reply:
All of them are access from `DatabaseConfig` and a couple of other places. I managed to make `firebaseApp` package-protected though.

Reply:
In addition to decorating the `ThreadFactory` this class also customizes the executor a little bit by calling `setRemoveOnCancelPolicy()`. I'd rather not to duplicate that in all the places we use this class. Also having all this in one place gives us an easy way to introduce more customizations in the future if required. WDYT?

Reply:
Another way to get around this would be to introduce a new util method somewhere:

```
public static ScheduledThreadPoolExecutor newScheduledExecutor(ThreadFactory factory, String name)
```

We can do all the decorating/customizing in the method. `DefaultRunLoop` will have to find another approach though, since it extends the executor. Thoughts?

Reply:
I like your last approach better, but I can't think of a good way to replace the inheritance in DefaultRunLoop. We might want to ponder a little more about this, but I don't think we should hold up this PR on it. If anything, we could argue that this derisks this change since it keeps the existing pattern.